### ISSUE TITLE
Semi- Supervised Technique ( Pseudo Labeling  or Active Learning ) With Yolov8 Segmentation .

### ISSUE THREAD

**Issue by AqsaM1** [ultralytics/ultralytics#14611]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi , I hope you are good.
I'm doing my master's thesis with my company and they need to generate ground truth for their large image data coming from scanner( redundant images). Previously external annotator was doing this step. Now in my thesis the proposed research is to use SAM(Segment Anything Model) to act as an external annotator and get segments for a small subset of dataset. As SAM is computationally expensive to get Segments from such a big dataset.

**Approach 1:**
So the pipeline goes like this small set (30 images) will be labeled by SAM then the it will be converted to yolo format and train yolo then make prediction from trained model filter the low confidence images(k=20) and give it to SAM . Get the segments and again convert it to yolo format and train yolo again this is iterative process. This is very slow and time consuming process
Currently I'm working with Stanford Indoor public dataset.

**Approach 2:**  

In 2nd approach I have integrated Active learning where I'm collecting random samples from Unlabeled Dataset and passing it to SAM and the output becomes the input for yolo.

I have used  model = YOLO(yolov8n-seg.yaml ).load('yolov8n.pt'). I tried to run pre-trained seg model with coco and also created data.yml file with coco classes. I'm not getting very significant improvement.

**These are my settings**
epochs:200
image_size=640
batch_size=4 
I have also use tune method to get the hyperparameters and override that. 
I have 12 labels in my dataset. 
The first three images are for random sampling - Active Learning
**This image is with coco classes integrated with data.yml file and used pretrained segmentation model**
![With_COCO_seg-pt](https://github.com/user-attachments/assets/a0d475a4-741c-4338-992e-cbed1336d014)

**This image is without  coco classes and used pretrained segmentation model**
![Without_COCO_Results-segpt](https://github.com/user-attachments/assets/2d7aa6fb-6c4b-4bd8-8121-064c7486998c)

**This image is without coco classes and used  segmentation model scratch and loaded detection model weight**
![Without_COCO_n_load_model](https://github.com/user-attachments/assets/c2f8b2c9-701f-4f10-962f-886277eff545)

**These are the current result I have trained my model using self training technique it's still running and currently I have 450 images in my training data.** 
![selftrain_yolo](https://github.com/user-attachments/assets/8be68699-01eb-4b03-b19c-1c4bdddf4e12)

I'm not able to pin point the issue, as box loss is decreasing but my segmentation loss and classification loss is mess. Training data could be one issue.  
Also I have one confusion about the conversion of Masks from SAM to yolo format I have written custom code for that . That could be an issue .  In label.txt file after class label it has to have bounding box coordinates or we can directly write the polygons coordinates?

 I need some suggestions to improve the performance of my pipeline.
Thank you

**Comment by Y-T-G** [comment]
> In label.txt file after class label it has to have bounding box coordinates or we can directly write the polygons coordinates?

Polygon.

Did you verify the labels in the train batch plots in the runs folder?

**Comment by Y-T-G** [comment]
> In 2nd approach I have integrated Active learning where I'm collecting random samples from Unlabeled Dataset and passing it to SAM and the output becomes the input for yolo.

In active learning, you use an actual human to label them. Using automatic annotation defeats the point of active learning. Active learning is about using small but highly informative and well labeled data to train a better model.

**Comment by AqsaM1** [comment]
> > In label.txt file after class label it has to have bounding box coordinates or we can directly write the polygons coordinates?
> 
> Polygon.
> 
> Did you verify the labels in the train batch plots in the runs folder?

I did but I'm 


> > In 2nd approach I have integrated Active learning where I'm collecting random samples from Unlabeled Dataset and passing it to SAM and the output becomes the input for yolo.
> 
> In active learning, you use an actual human to label them. Using automatic annotation defeats the point of active learning. Active learning is about using small but highly informative and well labeled data to train a better model.

If you see the masks generated by SAM, these are highly informative and accurate. There is one more block that I have integrated suggested by my manager that after getting segments from SAM one person will review it and delete the incorrect labels and pass it to training of yolo that set will be  small(20/30 images) to review so it wont be very time consuming. Very little human interaction.


![Image](https://github.com/user-attachments/assets/f523009d-368c-4fe9-a742-92d2e5f540ea)
![SAM](https://github.com/user-attachments/assets/ed8849fe-d9f2-4698-aeb6-045c14dfddd6)

**Comment by AqsaM1** [comment]
> > In label.txt file after class label it has to have bounding box coordinates or we can directly write the polygons coordinates?
> 
> Polygon.
> 
> Did you verify the labels in the train batch plots in the runs folder?

I did but I'm not very satisfied with that. I was going through a thread where it is mentioned for the segmentation task the label file should contain
class_id x y w h then polygons points. so I was thinking may be that could be the issue

**Comment by pderrenger** [comment]
Thank you for the detailed information and images. If your training batch plots show correct polygons, it indicates that your label conversion process is likely correct. For segmentation tasks, the label file should indeed contain polygon coordinates directly after the class ID. If you're still facing issues with segmentation loss, consider verifying the quality and consistency of your dataset annotations. Additionally, ensure that your training parameters are well-tuned for your specific dataset. If the problem persists, updating to the latest version of the Ultralytics package might help resolve any underlying issues.

**Comment by AqsaM1** [comment]
Hi @pderrenger  Thank you for replying, 
I have tried to tune my hyperparameter using the tune method. 40 epoch with 100 iteration adn used that pass these paramters while calling model.train(). By couldn't get significant result still my segmentation and classification loss is increasing. As my labeled set in small tune method is performing well. That could be the reason?

**Comment by Y-T-G** [comment]
I think the dataset is too small for training

**Comment by AqsaM1** [comment]
@Y-T-G 
Yes that I understand, but with the increase in training set, If you see the segmentation loss it is increasing , 
train set increase from 100 till 1k data-points.

**Comment by AqsaM1** [comment]
Hi, I'm working on stanford indoor dataset. Now I have trained yolov8 seg model on one of the area of stanford data which has around 10k images. and also tune my hyperparameters
Now model is giving good results on validation set and. I was following on of the threads to self train yolo, these are the steps mentioned by @glenn-jocher 

# Train a YOLOv8 model on a small set of labeled data.
# Use this model to generate predictions for a larger set of unlabeled data.
# Filter out low-confidence predictions based on a threshold. These are the predictions that are likely to be incorrect.
# Sort the remaining predictions by their confidence scores, and select the top k predictions, where k is the number of labels you want to collect in the next round.
# Manually label the selected predictions, and add them to your labeled data set.
# Train a new model on the combined labeled data set.
# Repeat steps 2-6 until the desired level of accuracy is reached, or until the model has been trained on all available data

There is one confusion I have, when saving the txt file after the prediction, it contains multiple class labels, polygons and the confidence score, If there are some classes which has higher confidence and some has lower then how to select the next batch of image to label..

I really need some guidance in that as I'm new to this domain. 
Thank you

**Comment by Y-T-G** [comment]
> For that I need probabilities to calculate the entropy, yolo predict func gives confidence score. How to approach this thing?

You will have to use the raw outputs and apply NMS manually.
`raw_output = model.predict(img, embed=[-1]) # (1, 84, 6300)`

**Comment by AqsaM1** [comment]
@Y-T-G  As I'm working on segmentation task, I have checked this raw_output but it is giving me 
list of len 2 and   0th index it has tensor (1,43, 8400) ( I have 10 classes )  and 1st index it has tuple. 
It there any documentation link available to check understand this raw output?

**Comment by AqsaM1** [comment]
Hi @pderrenger Could you please guide me about how can I implement dropout of L2 Regularization in yolov8seg model? Because of my small dataset, I'm facing overfitting in my loss curves and also alot of fluctuations.
Thank you

**Comment by pderrenger** [comment]
Hi @AqsaM1-G, you can implement dropout and L2 regularization in the YOLOv8 segmentation model by modifying the model's configuration file. For dropout, you can add a dropout layer in the model architecture. For L2 regularization, you can set the `weight_decay` parameter in the optimizer settings. Ensure you are using the latest version of the Ultralytics package to avoid any compatibility issues. If you need further details, please refer to the official documentation or the relevant sections in the model configuration files.

**Comment by AqsaM1** [comment]
> > For that I need probabilities to calculate the entropy, yolo predict func gives confidence score. How to approach this thing?
> 
> You will have to use the raw outputs and apply NMS manually. `raw_output = model.predict(img, embed=[-1]) # (1, 84, 6300)`
> 
> The 84 here are the 80 class probabilities and 4 bbox coordinates

hi @Laughing-q  I'm trying to implement entropy sampling to find the least confident sample. For the segmentation task I have checked this raw_output and it is giving me list of len 2 and 0th index it has tensor (1,48, 8400) ( I have 12 classes ) and 1st index it has tuple.
It there any documentation link available to check understand this raw output?

Can you guide me about how to approach this?
Thank you

**Comment by AqsaM1** [comment]
Hi @glenn-jocher  I need to understand about this raw output  as I'm using yolov8n-seg model for instance segmentation.
   
```
   raw_output = model.predict(images, embed=[-1])
   # 0th index tensor of shape (3, 48, 8400)
   #  1st index it is tuple of len 3
   Raw output shape 0:, raw_output[0].shape)- 3d tensor
   Raw output shape 1:" , len(raw_output[1])) - tuple
```

As I have three images in my images folder I assume, (3,48,8400) , 3 is number of images or batch_size , someone mentioned about the 2nd index contains the number of classes + bb coordinates . I have 12 classes in my dataset not sure about the remaining 36 value. What does 3rd index contain? and Why raw_output is a tuple of len 2 and what it contains on 2nd index of tuple?

#Help

**Comment by glenn-jocher** [comment]
Hi @AqsaM1, 

The raw output for the YOLOv8 segmentation model indeed includes a tensor and a tuple. The tensor shape (3, 48, 8400) indicates 3 images, 48 channels (12 classes * 4), and 8400 anchors. The tuple contains additional model-specific outputs. For detailed understanding, refer to the model's architecture documentation or source code.
